### ISSUE TITLE
make distcheck failing

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#842]
Just messing around and noticed make distcheck wasn't working.

First issue was easy to fix.

```
Making all in bindings
make[3]: Entering directory `/tmp/achu/flux-core/flux-core-0.4.1-217-g2df89fd/_build/src/bindings'
Making all in lua
make[4]: Entering directory `/tmp/achu/flux-core/flux-core-0.4.1-217-g2df89fd/_build/src/bindings/lua'
  CCLD     flux/affinity.la
libtool: link: `lua-affinity/lua-affinity.lo' is not a valid libtool object
```
Basically listing directories here in src/bindings/lua/Makefile.am:

```
 EXTRA_DIST = \
        README \
        $(TESTS) \
        Test \
        lua-affinity \
        lua-hostlist
```

leads to files being disted out twice b/c some files were listed earlier in the Makefile.

The second issue is quirkier.

```
make[2]: Entering directory `/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/src/bindings/python/flux'
make[2]: Nothing to be done for `install-exec-am'.
 /usr/bin/mkdir -p '/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_inst/lib64/python2.7/site-packages/flux'
 /bin/sh ../../../../libtool   --mode=install /usr/bin/install -c   _core.la _barrier.la _kvs.la _jsc.la _kz.la '/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_inst/lib64/python2.7/site-packages/flux'
libtool: install: warning: relinking `_core.la'
libtool: install: (cd /tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/src/bindings/python/flux; /bin/sh /tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/libtool  --silent --tag CC --mode=relink gcc -std=gnu99 -g -O2 -avoid-version -module -Wl,--no-undefined -L/usr/lib64 -lpython2.7 -Wl,-rpath,/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_inst/lib -o _core.la -rpath /tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_inst/lib64/python2.7/site-packages/flux _core.lo ../../../../src/common/libflux-core.la ../../../../src/common/libflux-internal.la -ljson-c -lczmq -lzmq -lm )
.libs/_core.o: In function `_cffi_f_flux_reactor_time':
/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/src/bindings/python/flux/_core.c:8586: undefined reference to `flux_reactor_time'
.libs/_core.o: In function `_cffi_f_flux_reactor_now_update':
/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/src/bindings/python/flux/_core.c:8440: undefined reference to `flux_reactor_now_update'
.libs/_core.o: In function `_cffi_f_flux_reactor_now':
/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/src/bindings/python/flux/_core.c:8405: undefined reference to `flux_reactor_now'
.libs/_core.o: In function `_cffi_d_flux_reactor_time':
/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/src/bindings/python/flux/_core.c:8576: undefined reference to `flux_reactor_time'
.libs/_core.o: In function `_cffi_d_flux_reactor_now_update':
/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/src/bindings/python/flux/_core.c:8418: undefined reference to `flux_reactor_now_update'
.libs/_core.o: In function `_cffi_d_flux_reactor_now':
/tmp/achu/flux-core/flux-core-0.4.1-218-g3f4593a/_build/src/bindings/python/flux/_core.c:8382: undefined reference to `flux_reactor_now'
collect2: error: ld returned 1 exit status
libtool: install: error: relink `_core.la' with the above command before installing it
make[2]: *** [install-fluxpyLTLIBRARIES] Error 1
```

the settings of AM_LDFLAGS in src/bindings/python/flux/Makefile.am are:

```
AM_LDFLAGS = \
	   -avoid-version -module $(san_ld_zdef_flag) \
	   $(PYTHON_LDFLAGS) -Wl,-rpath,$(PYTHON_PREFIX)/lib \
	   $(CODE_COVERAGE_LDFLAGS)
```

PYTHON_LDFLAGS = ```-L/usr/lib64 -lpython2.7```, so libflux-core.so is linked from /usr/lib64 instead of the project directory.  Unhappiness happens.

Other than outright add -L"some build dir path" I'm not entirely sure if there's a clean way to fix this b/c with PYTHON_LDFLAGS we're telling the linker to look in a standard library dir way earlier than it should.  Any thoughts?  

We of course can just leave it be ... found it so I thought I'd bring it up.

**Comment by garlick** [comment]
What environment?

**Comment by chu11** [comment]
This is on hype2.  Actually problem could be more general than on distcheck.  I did a basic:

```
./autogen.sh ; ./configure --prefix=/tmp/achu/fluxtry ; make; make install
```

and got the same problem.

```
libtool: install: warning: relinking `_core.la'
libtool: install: (cd /tmp/achu/flux-core/src/bindings/python/flux; /bin/sh /tmp/achu/flux-core/libtool  --silent --tag CC --mode=relink gcc -std=gnu99 -g -O2 -avoid-version -module -Wl,--no-undefined -L/usr/lib64 -lpython2.7 -Wl,-rpath,/tmp/achu/fluxtry/lib -o _core.la -rpath /tmp/achu/fluxtry/lib64/python2.7/site-packages/flux _core.lo ../../../../src/common/libflux-core.la ../../../../src/common/libflux-internal.la -ljson-c -lczmq -lzmq -lm )
.libs/_core.o: In function `_cffi_f_flux_reactor_time':
/tmp/achu/flux-core/src/bindings/python/flux/_core.c:8586: undefined reference to `flux_reactor_time'
.libs/_core.o: In function `_cffi_f_flux_reactor_now_update':
/tmp/achu/flux-core/src/bindings/python/flux/_core.c:8440: undefined reference to `flux_reactor_now_update'
.libs/_core.o: In function `_cffi_f_flux_reactor_now':
/tmp/achu/flux-core/src/bindings/python/flux/_core.c:8405: undefined reference to `flux_reactor_now'
.libs/_core.o: In function `_cffi_d_flux_reactor_time':
/tmp/achu/flux-core/src/bindings/python/flux/_core.c:8576: undefined reference to `flux_reactor_time'
.libs/_core.o: In function `_cffi_d_flux_reactor_now_update':
/tmp/achu/flux-core/src/bindings/python/flux/_core.c:8418: undefined reference to `flux_reactor_now_update'
.libs/_core.o: In function `_cffi_d_flux_reactor_now':
/tmp/achu/flux-core/src/bindings/python/flux/_core.c:8382: undefined reference to `flux_reactor_now'
collect2: error: ld returned 1 exit status
libtool: install: error: relink `_core.la' with the above command before installing it
```

**Comment by garlick** [comment]
FWIW we do run `make distcheck` in travis.

Lacking any real insight into what might be happening here, I'll throw out a couple of thoughts though you probably already have this covered:
* do a `make clean` and rebuild
* make sure lc local dotkit environment is disabled
* make sure `flux-core` module is loaded (or just its dependent modules)

**Comment by chu11** [comment]
I'll have to see what you guys do in your environment that's different than me.  For me, it seems if you use the default installed python (i.e. /usr/bin/python), then PYTHON_LDFLAGS will be set to "-L/usr/lib64" via configure, which makes sense.  But if Flux is also installed on the system in /usr/lib64 (i.e. /usr/lib64/libflux-core.so is there) it will get found/loaded by the linker before the one in the project b/c of PYTHON_LDFLAGS.  Some simple hackery (```find . -name Makefile | xargs sed -i -e "s/-L\/usr\/lib64 -lpython2.7/-lpython2.7/"```) gets around the problem.

**Comment by lipari** [comment]
... and whatever you discover as the cause should be added to [this page](http://flux-framework.github.io/docs/quickstart/) to the rest of the world doesn't get snagged on the burr you discovered.

**Comment by garlick** [comment]
Sorry - didn't realize you were on a toss3/rhel7 node of hype.  I'm building there now (very slowly due to NFS home dir probs - should have started in /tmp like you did)

Because it's TOSS3 the dotkit comment doesn't apply.  It's `/usr/tce` that should be avoided (touch `~/.notce`).  Not sure if that has any bearing here - it sounds like you've found a genuine problem with the way python is linking to libflux-core and are picking up the "correct" (system) python.

**Comment by grondo** [comment]
You definitely have to build without /usr/tce, see TOSS-3304 for *at least* one problem with the /usr/tce python (it can't import system python modules)

**Comment by garlick** [comment]
I did just hit the first `lua-affinity.lo` issue while doing `make distcheck` on hype2.

**Comment by chu11** [comment]
@garlick I hit the ```lua-affinity.lo``` this on a TOSS2 system as well.  Not sure if this is a subtlety of the autoconf/make versions.  My ugly fix was:

```patch
diff --git a/src/bindings/lua/Makefile.am b/src/bindings/lua/Makefile.am
index 0fe38fa..72f7f11 100644
--- a/src/bindings/lua/Makefile.am
+++ b/src/bindings/lua/Makefile.am
@@ -153,9 +153,22 @@ TESTS = \
 EXTRA_DIST = \
        README \
        $(TESTS) \
-       Test \
-       lua-affinity \
-       lua-hostlist
+       Test/COPYRIGHT \
+       Test/README.md \
+       Test/dist.info \
+       lua-affinity/COPYING \
+       lua-affinity/DISCLAIMER.LLNS \
+       lua-affinity/README.md \
+       lua-hostlist/COPYING \
+       lua-hostlist/DISCLAIMER \
+       lua-hostlist/META \
+       lua-hostlist/NEWS \
+       lua-hostlist/README.md \
+       lua-hostlist/test.lua \
+       lua-hostlist/tests/lunit \
+       lua-hostlist/tests/lunit-console.lua \
+       lua-hostlist/tests/lunit.lua \
+       lua-hostlist/tests/tests.lua
```

**Comment by garlick** [comment]
Just ran a `make distcheck` on my desktop and it failed on the lua problem too.  Maybe related to newer versions of autoconf than we have on travis?  Both hype2 and my desktop are 2.69 based.  Your patch fixes it.

I was doing this on my branch for pr #835 and hit this in distcheck:
```
make[5]: Nothing to be done for 'install-exec-am'.
 /bin/mkdir -p '/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_inst/lib/flux/modules'
 /bin/bash ../../../libtool   --mode=install /usr/bin/install -c   aggregator.la '/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_inst/lib/flux/modules'
libtool: warning: relinking 'aggregator.la'
libtool: install: (cd /home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_build/sub/src/modules/aggregator; /bin/bash "/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_build/sub/libtool"  --silent --tag CC --mode=relink gcc -Wall -Werror -Wno-strict-aliasing -Wno-error=deprecated-declarations -g -O2 -Wl,--no-undefined -avoid-version -export-symbols-regex "^mod_(main|name|service)\$" --disable-static -shared -export-dynamic -module -o aggregator.la -rpath /home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_inst/lib/flux/modules aggregator.lo ../../../src/common/libflux-internal.la ../../../src/common/libflux-core.la ../../../src/modules/kvs/libflux-kvs.la ../../../src/modules/barrier/libflux-barrier.la -L/usr/local/lib -lczmq -lzmq -lm )
.libs/aggregator.o: In function `aggregate_forward':
/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_build/sub/src/modules/aggregator/../../../../../src/modules/aggregator/aggregator.c:206: undefined reference to `flux_rpc_get_nodeid'
collect2: error: ld returned 1 exit status
libtool:   error: error: relink 'aggregator.la' with the above command before installing it
Makefile:495: recipe for target 'install-fluxmodLTLIBRARIES' failed
make[5]: *** [install-fluxmodLTLIBRARIES] Error 1
make[5]: Leaving directory '/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_build/sub/src/modules/aggregator'
Makefile:665: recipe for target 'install-am' failed
make[4]: *** [install-am] Error 2
make[4]: Leaving directory '/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_build/sub/src/modules/aggregator'
Makefile:462: recipe for target 'install-recursive' failed
make[3]: *** [install-recursive] Error 1
make[3]: Leaving directory '/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_build/sub/src/modules'
Makefile:467: recipe for target 'install-recursive' failed
make[2]: *** [install-recursive] Error 1
make[2]: Leaving directory '/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_build/sub/src'
Makefile:539: recipe for target 'install-recursive' failed
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory '/home/garlick/proj/flux-core/flux-core-0.4.1-244-gded4a18/_build/sub'
Makefile:745: recipe for target 'distcheck' failed
make: *** [distcheck] Error 1
```

so it would seem that we're linking against the older libflux-core.so, maybe the one I have installed in /usr/local/lib, since the error is a missing function that I just added in this PR.

**Comment by grondo** [comment]
That is pretty ugly. In Travis we're always building from a "clean" system without a flux-core installed so that could be allowing `make distcheck` to pass more easily there as well.

**Comment by chu11** [comment]
Yeah, it seems to be a combination of libflux-core and python being installed in standard locations.  Setting ```PYTHON_LDFLAGS="-lpython2.7"``` at configure time is more than sufficient to work around the issue on systems with the issue, but I haven't been able to come up with a nice solution.  Everything that works more or less involves just getting rid of ```-L/usr/lib64``` as a special case.

**Comment by chu11** [comment]
I thought of this on the way home tonight and came up with something.  I put the change in ```ax_python_devel.m4``` which is less than ideal.  It's not perfect, but I don't think it's horrific.

```patch
diff --git a/config/ax_python_devel.m4 b/config/ax_python_devel.m4
index 59a2ff0..f1d0126 100644
--- a/config/ax_python_devel.m4
+++ b/config/ax_python_devel.m4
@@ -235,6 +235,7 @@ EOD`
                          import os; \
                          print (os.path.join(f(plat_specific=1, standard_lib=1), 'config'));"`
                        PYTHON_LDFLAGS="-L$ac_python_libdir -lpython$ac_python_version"
+                       ac_python_library="-lpython$ac_python_version"
                fi
 
                if test -z "PYTHON_LDFLAGS"; then
@@ -247,6 +248,16 @@ EOD`
        AC_MSG_RESULT([$PYTHON_LDFLAGS])
        AC_SUBST([PYTHON_LDFLAGS])
 
+       # Remove -L<path> from PYTHON_LDFLAGS if it is in a standard path,
+       # including standard path earlier in LD search can lead to problems
+       AC_CHECK_LIB([$ac_python_library], [PyArg_ParseTuple], 
+                    [ac_python_in_ld_path=true])
+       if test "$ac_python_in_ld_path" = "true"; then
+               AC_MSG_NOTICE([Removing -L$ac_python_libdir from PYTHON_LDFLAGS])
+               PYTHON_LDFLAGS="-l$ac_python_library"
+       fi
+       AC_SUBST([PYTHON_LDFLAGS])
+
        #
        # Check for site packages
        #
```

**Comment by garlick** [comment]
If that's the only source of the extraneous -L then it looks good to me.  Want to propose a PR for this and the EXTRA_DIST fix?

**Comment by chu11** [comment]
Yup, I'll clean it up and send a PR.

**Comment by grondo** [comment]
Did `ax_python_devel.m4` come from [autoconf-archive](https://www.gnu.org/software/autoconf-archive/ax_python_devel.html)? If so, maybe the w/a should go directly in `configure.ac` instead?

**Comment by chu11** [comment]
Yeah, which is why I didn't like my patch too much.  But I think I can sneak this into ```configure.ac``` because we have a requirement of atleast python 2.7.  So I don't really need to handle the old python case.

**Comment by chu11** [comment]
fixed in #847
